### PR TITLE
feat: support set global variable

### PR DESCRIPTION
**Pull request by keyu813**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature: 
support set@@global.xxx = xxx & set GLOBAL xxx = xxx for #1257 

企业微信截图_8b8e5800-c06d-4f23-9b36-23b0cedc0203

### CODE REVIEW

File: src/sdk/sql_cluster_router.cc [hunk 1]
```diff
@@ -2315,11 +2315,32 @@ bool SQLClusterRouter::IsEnableTrace() {
 }
 
 ::hybridse::sdk::Status SQLClusterRouter::SetVariable(hybridse::node::SetPlanNode* node) {
-    if (node->Scope() == hybridse::node::VariableScope::kGlobalSystemVariable) {
-        return {::hybridse::common::StatusCode::kCmdError, "global system variable is unsupported"};
-    }
     std::string key = node->Key();
     std::transform(key.begin(), key.end(), key.begin(), ::tolower);
+    if (node->Scope() == hybridse::node::VariableScope::kGlobalSystemVariable) {
+        std::lock_guard<::openmldb::base::SpinMutex> lock(mu_);
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc]: why add lock

File: src/sdk/sql_cluster_router.cc [hunk 2]
```diff
@@ -2315,11 +2315,32 @@ bool SQLClusterRouter::IsEnableTrace() {
...
+        std::string value = node->Value()->GetExprString();
+        std::transform(value.begin(), value.end(), value.begin(), ::tolower);
+        if (key != "execute_mode" && key != "enable_trace") {
+            return {::hybridse::common::StatusCode::kCmdError, "no global variable " + key};
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc]: `execute_mode` and `enable_trace` are session variables

File: src/sdk/sql_cluster_router.cc [hunk 3]
```diff
@@ -2315,11 +2315,32 @@ bool SQLClusterRouter::IsEnableTrace() {
...
+        if (key == "execute_mode") {
+            if (value != "online" && value != "offline") {
+                return {::hybridse::common::StatusCode::kCmdError, "the value of execute_mode must be online|offline"};
+            }
+        } else if (key == "enable_trace") {
+            if (value != "true" && value != "false") {
+                return {::hybridse::common::StatusCode::kCmdError, "the value of enable_trace must be true|false"};
+            }
+        }
+        hybridse::sdk::Status status;
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc:2419]: use the output parameter `status` of `ExecuteSQL`

File: src/sdk/sql_cluster_router.cc [hunk 4]
```diff
@@ -2315,11 +2315,32 @@ bool SQLClusterRouter::IsEnableTrace() {
...
+        std::string sql = "INSERT INTO GLOBAL_VARIABLES values(" + key + ", " + value + ");";
+        ExecuteInsert("INFORMATION_SCHEMA", sql, &status);
+        if (status.code != 0) {
+            std::cout << "ERROR:" << status.msg << std::endl;
```
Comment (dl239) [inline src/sdk/sql_cluster_router.cc]: do not print msg

File: src/sdk/sql_cluster_router.cc (hunk 2 above)
Comment (keyu813) [reply src/sdk/sql_cluster_router.cc]: > `execute_mode` and `enable_trace` are session variables

Actually I have no idea now what global variables we have now...

File: src/cmd/sql_cmd_test.cc [hunk 5]
```diff
@@ -309,13 +309,6 @@ TEST_P(DBSDKTest, create_without_index_col) {
     ASSERT_TRUE(cs->GetNsClient()->DropTable("test2", "trans", msg));
 }
 
-TEST_P(DBSDKTest, show) {
```
Comment (aceforeverd) [inline src/cmd/sql_cmd_test.cc]: why rm this case ?

Comment (keyu813) [reply src/cmd/sql_cmd_test.cc]: not rm, cases about global variable are removed into sql_cluster_test

File: src/sdk/db_sdk.cc [hunk 6]
```diff
@@ -40,6 +40,7 @@ ClusterSDK::ClusterSDK(const ClusterOptions& options)
       table_root_path_(options.zk_path + "/table/db_table_data"),
       sp_root_path_(options.zk_path + "/store_procedure/db_sp_data"),
       notify_path_(options.zk_path + "/table/notify"),
+      globalvar_changed_notify_path_(options.zk_path + "/global_variable/changed"),
```
Comment (dl239) [inline src/sdk/db_sdk.cc]: change to /notify/global_variable

File: src/sdk/sql_cluster_test.cc [hunk 7]
```diff
@@ -845,6 +845,34 @@ TEST_F(SQLClusterTest, GetTableSchema) {
     ASSERT_TRUE(router->ExecuteDDL(db, "drop table test0;", &status));
     ASSERT_TRUE(router->DropDB(db, &status));
 }
+
+TEST_F(SQLClusterTest, GlobalVariable) {
```
Comment (aceforeverd) [inline src/sdk/sql_cluster_test.cc]: there is no test for standalone mode

File: src/tablet/tablet_impl.cc [hunk 8]
```diff
@@ -3795,6 +3807,27 @@ bool TabletImpl::RefreshSingleTable(uint32_t tid) {
...
+    if (status.code != 0) {
+        LOG(ERROR) << "update global var table failed: " << status.msg;
+        return;
+    }
+    std::string key;
+    std::string value;
+    while (rs->Next()) {
+        key = rs->GetStringUnsafe(0);
+        value = rs->GetStringUnsafe(1);
+        global_variables_[key] = value;
```
Comment (dl239) [inline src/tablet/tablet_impl.cc]: guarantee threadsafe

File: src/sdk/sql_cluster_test.cc [hunk 9]
```diff
@@ -845,6 +845,34 @@ TEST_F(SQLClusterTest, GetTableSchema) {
...
+        ns_client->ShowTable(nameserver::GLOBAL_VARIABLE_NAME, nameserver::INFORMATION_SCHEMA_DB, false, tables, msg));
+    ASSERT_EQ(1, tables.size());
+    ASSERT_STREQ(nameserver::GLOBAL_VARIABLE_NAME, tables[0].name().c_str());
+    tables.clear();
+
+    ::hybridse::sdk::Status status;
+    std::string sql = "set @@global.enable_trace='true';";
+    auto res = router->ExecuteSQL(sql, &status);
+    ASSERT_EQ(0, status.code);
+    ASSERT_TRUE(router->ExecuteSQL("show global variables", &status));
```
Comment (aceforeverd) [inline src/sdk/sql_cluster_test.cc]: should assert the output rows

File: src/sdk/sql_cluster_router.cc [hunk 10]
```diff
@@ -2412,11 +2412,33 @@ bool SQLClusterRouter::IsEnableTrace() {
...
 ::hybridse::sdk::Status SQLClusterRouter::SetVariable(hybridse::node::SetPlanNode* node) {
-    if (node->Scope() == hybridse::node::VariableScope::kGlobalSystemVariable) {
-        return {::hybridse::common::StatusCode::kCmdError, "global system variable is unsupported"};
-    }
     std::string key = node->Key();
     std::transform(key.begin(), key.end(), key.begin(), ::tolower);
+    if (node->Scope() == hybridse::node::VariableScope::kGlobalSystemVariable) {
+        std::string value = node->Value()->GetExprString();
+        std::transform(value.begin(), value.end(), value.begin(), ::tolower);
+        if (key != "execute_mode" && key != "enable_trace") {
```
Comment (aceforeverd) [inline src/sdk/sql_cluster_router.cc]: better not limit what the global key is, keep it just as a insert stmt

File: src/sdk/sql_cluster_test.cc [hunk 11]
```diff
@@ -845,6 +845,34 @@ TEST_F(SQLClusterTest, GetTableSchema) {
...
+    ASSERT_EQ(1, tables.size());
+    tables.clear();
+    ASSERT_TRUE(
+        ns_client->ShowTable(nameserver::GLOBAL_VARIABLE_NAME, nameserver::INFORMATION_SCHEMA_DB, false, tables, msg));
+    ASSERT_EQ(1, tables.size());
+    ASSERT_STREQ(nameserver::GLOBAL_VARIABLE_NAME, tables[0].name().c_str());
+    tables.clear();
+
+    ::hybridse::sdk::Status status;
+    std::string sql = "set @@global.enable_trace='true';";
```
Comment (aceforeverd) [inline src/sdk/sql_cluster_test.cc]: test `SET GLOBAL` as well

**Review by aceforeverd — CHANGES_REQUESTED** [review]

**Comment by dl239** [comment]
PR review meeting note:
- Modify notify path
- Rename the `db_`
- Recover the global variables after `tablet` startup
- Init  global/session variables from system table in `sdk`

**Comment by dl239** [comment]
PR review meeting note:
- Modify notify path
- Rename the `db_`
- Recover the global variables after `tablet` startup
- Init  global/session variables from system table in `sdk`

File: src/sdk/sql_cluster_test.cc (hunk 7 above)
Comment (keyu813) [reply src/sdk/sql_cluster_test.cc]: create system do not support standalone mode now
企业微信截图_83cdc6fe-9b1c-463d-ad09-d25c70c56c29

File: src/sdk/db_sdk.h [hunk 12]
```diff
@@ -202,6 +205,8 @@ class StandAloneSDK : public DBSDK {
 
     int GetPort() const { return port_; }
 
+    bool GlobalVarNotify() const override { return false; }
```
Comment (aceforeverd) [inline src/sdk/db_sdk.h]: return true

File: src/sdk/sql_cluster_test.cc (hunk 7 above)
Comment (aceforeverd) [reply src/sdk/sql_cluster_test.cc]: #1458

**Review by aceforeverd — APPROVED** [review]

File: src/sdk/db_sdk.cc [hunk 13]
```diff
@@ -96,6 +97,11 @@ bool ClusterSDK::TriggerNotify() const {
     return zk_client_->Increment(notify_path_);
 }
 
+bool ClusterSDK::GlobalVarNotify() const {
+    LOG(INFO) << "Global variable changed trigger notify node";
```
Comment (dl239) [inline src/sdk/db_sdk.cc:101]: Remove or set as DLOG

File: src/sdk/db_sdk.h [hunk 14]
```diff
@@ -123,6 +123,7 @@ class DBSDK {
                                                                    std::string* msg);
     std::vector<std::shared_ptr<hybridse::sdk::ProcedureInfo>> GetProcedureInfo(std::string* msg);
     virtual bool TriggerNotify() const = 0;
+    virtual bool GlobalVarNotify() const = 0;
```
Comment (dl239) [inline src/sdk/db_sdk.h:126]: Call TriggerNotify as different notify_type. can do later

**Review by dl239 — APPROVED** [review]

**Comment by keyu813** [comment]
[global variable support](https://trello.com/c/Nqn4zAcQ/29-global-variable-support)

**Pull request merged**